jquense/yup
# Add currentValue to schema

**New Feature**

When using yup with libraries like Formik or React Hook Forms in React, the entire yup schema is revalidated whenever an input is filled. This becomes problematic when server-side validation needs to be added, as every time a form input is typed, the whole schema is validated, triggering an API request. For instance, typing in the password field might cause the test function for the email field to run. 

To solve this issue, consider adding the current value to the schema for comparing it with the new value. If they are the same, the test function is not run, resulting in more efficient validation.

This is not a yup issue, only a new feature to integrate with existing javascript form validation libraries that use yup and require more control for server-side validations.

**Schema example:**

```ts
const validationSchema = yup.object({
  email: yup
    .string()
    .email()
    .required('Email is required')
    .test('email-exists', 'Email Validation Error', (value, ctx) => {
      if (value !== ctx.currentValue) { //<=== New feature added ctx.currentValue
        return new Promise(async (resolve) => {
          const response = await axios.get(`http://localhost:3000/users?email=${value}`);
          resolve(response.data.length ? ctx.createError({ message: 'Email Exists' }) : true);
        });
      }
    }),
  password: yup.string().min(8, 'Password should be of minimum 8 characters length').required('Password is required'),
});
```
`ctx.currentValue` refers to the current value the field has in the form, which will be compared with `value`, which is the new user input. `ctx.currentValue` is different than `ctx.originalValue`, this last one is the non-casted `value`.

**Steps for testing this current branch:**

1. Download the following example repository:
https://github.com/webblocksapp/react-formik-yup-server-validation

2. Compile the yup branch.
3. Inside the downloaded repo, run `pnpm link <yup-repository-dir>`
4. Run `pnpm dev` and `pnpm dev-server` to visualize the form.
5. Type the password field, you will see the console logs running for the email field.

**Expected behavior**

You will see on the repo example that the `.test` function content for the `email` field, will run once
`console.log('Email field test logic is executed')` when the password field is input.
